<a href="https://colab.research.google.com/github/Pimchanok6430205105/BSC_DPDM23/blob/main/%E0%B8%AA%E0%B8%B3%E0%B9%80%E0%B8%99%E0%B8%B2%E0%B8%82%E0%B8%AD%E0%B8%87_Frequent_Patterns_(Association_Rules).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
import pandas as pd
import os

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [4]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [13]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/BSC_DPDM23/reduced_marketbasket.csv'),encoding='latin-1')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [4]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

##แปลง data ให้อยู่ในรูปแบบ Transaction data ด้วย Groupby.apply(list)

In [5]:
data.groupby('InvoiceNo')['StockCode'].apply(list)

InvoiceNo
536365     [85123A, 71053, 84406B, 84029G, 84029E, 22752,...
536366                                        [22633, 22632]
536367     [84879, 22745, 22748, 22749, 22310, 84969, 226...
536368                          [22960, 22913, 22912, 22914]
536369                                               [21756]
                                 ...                        
C581484                                              [23843]
C581490                                       [22178, 23144]
C581499                                                  [M]
C581568                                              [21258]
C581569                                       [84978, 20979]
Name: StockCode, Length: 25900, dtype: object

In [6]:
transactions = data.groupby('InvoiceNo')['StockCode'].apply(list)

In [7]:
data.shape

(541909, 8)

###Quiz หาว่ามีกี่ประเทศ

In [8]:
data['Country'].value_counts().shape

(38,)

## Association Rules

In [9]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5954 sha256=b4a296898e6c79b879c6a0c8c0f5540895013e439f5d2604a58e136657c54c4e
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [10]:
from apyori import apriori

In [11]:
frequentItemsets = list(apriori(transactions,min_support = 0.065))
frequentItemsets

[RelationRecord(items=frozenset({'22423'}), support=0.08386100386100385, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22423'}), confidence=0.08386100386100385, lift=1.0)]),
 RelationRecord(items=frozenset({'47566'}), support=0.06586872586872587, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'47566'}), confidence=0.06586872586872587, lift=1.0)]),
 RelationRecord(items=frozenset({'85099B'}), support=0.08243243243243244, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B'}), confidence=0.08243243243243244, lift=1.0)]),
 RelationRecord(items=frozenset({'85123A'}), support=0.08671814671814672, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85123A'}), confidence=0.08671814671814672, lift=1.0)])]

In [12]:
len(frequentItemsets)

4

In [13]:
frequentItemsets[0][0]

frozenset({'22423'})

In [14]:
dir(frequentItemsets[0][0])

['__and__',
 '__class__',
 '__class_getitem__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__ne__',
 '__new__',
 '__or__',
 '__rand__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__ror__',
 '__rsub__',
 '__rxor__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__sub__',
 '__subclasshook__',
 '__xor__',
 'copy',
 'difference',
 'intersection',
 'isdisjoint',
 'issubset',
 'issuperset',
 'symmetric_difference',
 'union']

In [15]:
frequentItemsets[0][0].__len__()

1

In [16]:
min_sup = 0.03

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

yes
RelationRecord(items=frozenset({'85099B', '22386'}), support=0.03216216216216216, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'85099B', '22386'}), confidence=0.03216216216216216, lift=1.0), OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335), OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])
yes
RelationRecord(items=frozenset({'22697', '22699'}), support=0.03027027027027027, ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22697', '22699'}), confidence=0.03027027027027027, lift=1.0), OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947), OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])


RelationRecord(items=frozenset({'22386', '85099B'}),

support=0.03216216216216216,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22386', '85099B'}), confidence=0.03216216216216216, lift=1.0),

OrderedStatistic(items_base=frozenset({'22386'}), items_add=frozenset({'85099B'}), confidence=0.6766856214459789, lift=8.20897311262335),

OrderedStatistic(items_base=frozenset({'85099B'}), items_add=frozenset({'22386'}), confidence=0.39016393442622943, lift=8.20897311262335)])


In [17]:
set(data[(data['StockCode']=='22386') | (data['StockCode']=='85099B') | (data['StockCode']=='22699') | (data['StockCode']=='22697') ]['Description'])

{'GREEN REGENCY TEACUP AND SAUCER',
 'JUMBO BAG PINK POLKADOT',
 'JUMBO BAG RED RETROSPOT',
 'ROSES REGENCY TEACUP AND SAUCER '}

RelationRecord(items=frozenset({'22699', '22697'}), support=0.03027027027027027,

ordered_statistics=[OrderedStatistic(items_base=frozenset(), items_add=frozenset({'22699', '22697'}), confidence=0.03027027027027027, lift=1.0),

OrderedStatistic(items_base=frozenset({'22697'}), items_add=frozenset({'22699'}), confidence=0.7417218543046358, lift=17.1523178807947),

OrderedStatistic(items_base=frozenset({'22699'}), items_add=frozenset({'22697'}), confidence=0.7, lift=17.1523178807947)])

In [18]:
set(data[(data['StockCode']=='22699') | (data['StockCode']=='22697')]['InvoiceNo'])

{'562717',
 '543045',
 '564817',
 '561637',
 '566193',
 '568060',
 '564092',
 '551997',
 '558862',
 '571713',
 '537240',
 '542619',
 '567293',
 '537396',
 '579187',
 '549222',
 '564835',
 '555722',
 '546136',
 '558120',
 'C539071',
 '578817',
 '575703',
 '578067',
 '576202',
 '558116',
 '546660',
 '571034',
 '539594',
 '539609',
 '548987',
 '560246',
 'C545033',
 '545428',
 '549951',
 '560473',
 'C545438',
 '546896',
 '556206',
 '562711',
 '553891',
 '540499',
 '543107',
 '573585',
 '553035',
 '555851',
 '562282',
 '565150',
 '541960',
 '545411',
 '564191',
 '554013',
 '541584',
 '545216',
 '553542',
 'C579945',
 '563166',
 '545895',
 '552956',
 '559055',
 '538566',
 '550986',
 '562607',
 '551387',
 '548260',
 '542349',
 '569834',
 '567671',
 '552867',
 '571061',
 '572854',
 '538370',
 '559817',
 '552563',
 '545214',
 '552694',
 '561891',
 '539735',
 '544660',
 '543989',
 '573020',
 '576246',
 '537642',
 '555106',
 '573388',
 'C555409',
 '560826',
 '564057',
 '537179',
 '565396',
 '564

In [19]:
data[data['InvoiceNo']=='562950']

InvoiceNo StockCode                         Description  Quantity  \
297488    562950     22752        SET 7 BABUSHKA NESTING BOXES         2   
297489    562950     22755     SMALL PURPLE BABUSHKA NOTEBOOK         12   
297490    562950     22753     SMALL YELLOW BABUSHKA NOTEBOOK         12   
297491    562950     22754        SMALL RED BABUSHKA NOTEBOOK         12   
297492    562950     22757        LARGE RED BABUSHKA NOTEBOOK         12   
297493    562950     22758    LARGE PURPLE BABUSHKA NOTEBOOK          12   
297494    562950     22756     LARGE YELLOW BABUSHKA NOTEBOOK         12   
297495    562950     23310             BUBBLEGUM RING ASSORTED        36   
297496    562950     23280  FOLDING BUTTERFLY MIRROR HOT PINK         12   
297497    562950     22383              LUNCH BAG SUKI DESIGN         10   
297498    562950     22492             MINI PAINT SET VINTAGE         36   
297499    562950     71459      HANGING JAM JAR T-LIGHT HOLDER        12   
297500    562950     84945  MULTI COLOUR SILVER T-LIGHT HOLDER        12   
297501    562950     21175         GIN + TONIC DIET METAL SIGN        12   
297502    562950     23164                  REGENCY CAKE SLICE         4   
297503    562950     23245          SET OF 3 REGENCY CAKE TINS         4   
297504    562950     23171            REGENCY TEA PLATE GREEN         12   
297505    562950     22697     GREEN REGENCY TEACUP AND SAUCER         6   

             InvoiceDate  UnitPrice  CustomerID         Country  
297488  08/11/2011 09:56       8.50     15608.0  United Kingdom  
297489  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297490  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297491  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297492  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297493  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297494  08/11/2011 09:56       1.25     15608.0  United Kingdom  
297495  08/11/2011 09:56       0.42     15608.0  United Kingdom  
297496  08/11/2011 09:56       0.83     15608.0  United Kingdom  
297497  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297498  08/11/2011 09:56       0.65     15608.0  United Kingdom  
297499  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297500  08/11/2011 09:56       0.85     15608.0  United Kingdom  
297501  08/11/2011 09:56       2.55     15608.0  United Kingdom  
297502  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297503  08/11/2011 09:56       4.95     15608.0  United Kingdom  
297504  08/11/2011 09:56       1.65     15608.0  United Kingdom  
297505  08/11/2011 09:56       2.95     15608.0  United Kingdom

### Quiz 22386 และ 85099B คือสินค้าอะไร

In [20]:
set(data[data['StockCode']=='22386']['Description'])

{'JUMBO BAG PINK POLKADOT'}

In [21]:
set(data[data['StockCode']=='85099B']['Description'])

{'JUMBO BAG RED RETROSPOT'}

#Homework1 หา Frequent Item Set และ Assosiation Rules
-  ของช่วงเวลา Chrismas & New year (1 Dec-31 Dec) [หนามเตย]
-  ของประเทศในทวีป Europe [จจ]
-  ของช่วงเวลา ตรุษจีนในทวีป Asia(15 Jan-15 Feb) [ใบเตย]
-  ของช่วงเวลา Chrismas & New year ในทวีป Europe (1 Dec-31 Dec) [พิม]
-  ของวันวาเลนไทน์ [ตุนนี่]
-  โดยจัดกลุ่มตามประเทศ+เดือน [พิมจ๊อก]

In [22]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

             InvoiceDate  UnitPrice  CustomerID         Country  
0       12/01/2010 08:26       2.55     17850.0  United Kingdom  
1       12/01/2010 08:26       3.39     17850.0  United Kingdom  
2       12/01/2010 08:26       2.75     17850.0  United Kingdom  
3       12/01/2010 08:26       3.39     17850.0  United Kingdom  
4       12/01/2010 08:26       3.39     17850.0  United Kingdom  
...                  ...        ...         ...             ...  
541904  12/09/2011 12:50       0.85     12680.0          France  
541905  12/09/2011 12:50       2.10     12680.0          France  
541906  12/09/2011 12:50       4.15     12680.0          France  
541907  12/09/2011 12:50       4.15     12680.0          France  
541908  12/09/2011 12:50       4.95     12680.0          France  

[541909 rows x 8 columns]

##จัดการข้อมูล

แยกวัน เดือน ปี เวลา

In [14]:
import pandas as pd

data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

# Create new columns for day, month, year, and time
data['Day'] = data['InvoiceDate'].dt.day
data['Month'] = data['InvoiceDate'].dt.month
data['Year'] = data['InvoiceDate'].dt.year
data['Time'] = data['InvoiceDate'].dt.time

# Display the modified DataFrame
print(data.head())

# Drop the original
data.drop(columns=['InvoiceDate'], inplace=True)


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


  InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  Day  Month  \
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom    1     12   
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom    1     12   
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom    1     12   

   Year      Time  
0  2010  08:26:00  
1  2010  08:26:00  
2  2010  08:26:00  
3  2010  08:26:00  
4  2010  0

##1.ของช่วงเวลา Chrismas & New year (1 Dec-31 Dec)

In [24]:
import pandas as pd
import os
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules

In [25]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/Mining'

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [26]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/Mining/reduced_marketbasket.csv'),encoding='latin-1')

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


FileNotFoundError: [Errno 2] No such file or directory: '/content/drive/MyDrive/Mining/reduced_marketbasket.csv'

In [ ]:
data

In [ ]:
# แปลงคอลัมน์ timestamp เป็น datetime
data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'])

In [ ]:
# กรองข้อมูลตามช่วงเวลา Chrismas & New Year
start_date = '2010-12-01'
end_date = '2011-12-09'
data = data[(data['InvoiceDate'] >= start_date) & (data['InvoiceDate'] <= end_date)]


In [ ]:
# แยกวันที่เป็นเดือน, วัน, และปีแยกกัน
data['month'] = data['InvoiceDate'].dt.month
data['day'] = data['InvoiceDate'].dt.day
data['year'] = data['InvoiceDate'].dt.year


In [ ]:
#เช็ค
data[['InvoiceDate', 'month', 'day', 'year']].head()

In [ ]:
# แปลงข้อมูลเป็น Transaction List
transactions= data.groupby('InvoiceNo')['StockCode'].apply(list)

In [ ]:
# ใช้ TransactionEncoder เพื่อแปลง Transaction List เป็นรูปแบบที่ mlxtend สามารถใช้ได้
te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)
df = pd.DataFrame(te_ary, columns=te.columns_)

In [ ]:
# หา Frequent Item Set ด้วย apriori algorithm
frequent_itemsets = apriori(df, min_support=0.04, use_colnames=True)

In [ ]:
 #หา Association Rules
rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.5)

In [ ]:
# แสดง Frequent Item Set
print("Frequent Item Set:")
print(frequent_itemsets)

In [ ]:
# แสดง Association Rules
print("\nAssociation Rules:")
print(rules)

In [ ]:
data[data['StockCode']=='POST']

#ยังบ่เสร็จ

##2.ของประเทศในทวีป Europe

In [ ]:
import pandas as pd
import os

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

In [ ]:
data = pd.read_csv(os.path.join(path,'reduced_marketbasket.csv'),encoding='latin-1')

In [ ]:
data

In [ ]:
import pandas as pd

data['InvoiceDate'] = pd.to_datetime(data['InvoiceDate'], format='%m/%d/%Y %H:%M')

# Create new columns for day, month, year, and time
data['Day'] = data['InvoiceDate'].dt.day
data['Month'] = data['InvoiceDate'].dt.month
data['Year'] = data['InvoiceDate'].dt.year
data['Time'] = data['InvoiceDate'].dt.time

# Display the modified DataFrame
print(data.head())

# Drop the original
data.drop(columns=['InvoiceDate'], inplace=True)

In [ ]:
set(data['Country'])

In [ ]:
data_EU_country = ['Austria','Belgium','Cyprus','Channel Islands','Czech Republic','Denmark','EIRE','European Community',
              'Finland','France','Germany','Greece','Iceland','Italy','Lithuania','Malta','Netherlands',
              'Norway','Poland','Portugal','Spain','Sweden','Switzerland','United Kingdom']

In [ ]:
dataEU = data[data['Country']==""]
for i in range(len(data_EU_country)):
  x = data[data['Country']==data_EU_country[i]]
  dataEU=pd.concat([x,dataEU])

In [ ]:
len(set(dataEU['Country']))

In [ ]:
dataEU

In [ ]:
dataEU.groupby('InvoiceNo')['StockCode'].apply(list)


In [ ]:
transactions = dataEU.groupby('InvoiceNo')['StockCode'].apply(list)


In [ ]:
!pip install apyori

In [ ]:
from apyori import apriori


In [ ]:
apriori(transactions,min_support = 0.053)


In [ ]:
list(apriori(transactions,min_support = 0.053))


In [ ]:
frequentItemsets = list(apriori(transactions,min_support = 0.053))
frequentItemsets

In [ ]:
len(frequentItemsets)


In [ ]:
frequentItemsets[0]


In [ ]:
frequentItemsets[0][0]

In [ ]:
dir(frequentItemsets[0][0])

In [ ]:
frequentItemsets[0][0].__len__()


In [ ]:
min_sup = 0.032

frequentItemsets = list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

In [ ]:
set(dataEU[(dataEU['StockCode']=='85099B')]['Description'])


In [ ]:
set(dataEU[(dataEU['StockCode']=='22807')]['Description'])


In [ ]:
transactions =dataEU.groupby(['Country', 'Year'])['StockCode'].apply(list)
min_sup = 0.4
frequentItemsets= list(apriori(transactions,min_support = min_sup))
for i in frequentItemsets:
    if i[0].__len__()>1:
        print('yes')
        print(i)

In [ ]:
set(dataEU[(dataEU['StockCode']=='22423')]['Year'])


##3.ของช่วงเวลา ตรุษจีนในทวีป Asia(15 Jan-15 Feb)

In [ ]:
import pandas as pd
import os

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
path = '/content/drive/MyDrive/BSC_DPDM23'

In [ ]:
data = pd.read_csv(os.path.join(path,'/content/drive/MyDrive/BSC_DPDM23/reduced_marketbasket.csv'),encoding='latin-1')

In [ ]:
data

In [ ]:
Asia = data[(data['Country']=='Bahrain') | (data['Country']=='Hong Kong')| (data['Country']=='Singapore') | (data['Country']=='Japan') | (data['Country']=='Israel') |
            (data['Country']=='Lebanon') | (data['Country']=='Saudi Arabia') | (data['Country']=='United Arab Emirates')]
Asia #แบ่งทวีปเอเชีย


In [ ]:
# สร้าง DataFrame สำหรับเหตุการณ์ Chinese New Year
China = Asia[((Asia['Month'] == 1) & (Asia['Day'] >= 15) & (Asia['Day'] <= 31)) |
            ((Asia['Month'] == 2) & (Asia['Day'] <= 15))]

print(China)

In [ ]:
trans_china = China.groupby('InvoiceNo')['Description'].apply(list).reset_index()
trans_china_apply = trans_chin['Description'].apply(lambda x: [str(item) for item in x])

In [ ]:
trans_china_apply

In [ ]:
min_sup = 0.33 #เปลี่ยนจนกว่าจะเจอ...

frequentItemset_China = list(apriori(trans_chin_apply, min_support = min_sup))
for i in frequentItemset_China:
    if i[0].__len__()>1:
        print('yes')
        print(i)

In [ ]:
frequentItemset_China = list(apriori(trans_chin_apply, min_support=0.33))

for record in frequentItemset_China:
    if len(record.items) > 1:
        print('yes')
        print(f"Items: {', '.join(record.items)}")
        print(f"Support: {record.support:.4f} ({record.support*100:.2f}%)")

        for i, stat in enumerate(record.ordered_statistics, 1):
            items_base = 'None' if not stat.items_base else ', '.join(stat.items_base)
            items_add = ', '.join(stat.items_add)
            confidence = stat.confidence
            lift = stat.lift

            print(f"Rule {i}: ({items_base} -> {items_add} : Confidence={confidence:.4f}, Lift={lift:.4f})")
        print()

##4.ของช่วงเวลา Chrismas & New year ในทวีป Europe (1 Dec-31 Dec)

In [15]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

# Filter data for the Christmas & New Year period and European countries
european_countries = ['Austria', 'Belgium', 'Cyprus', 'Channel Islands', 'Czech Republic', 'Denmark', 'EIRE', 'European Community',
                      'Finland', 'France', 'Germany', 'Greece', 'Iceland', 'Italy', 'Lithuania', 'Malta', 'Netherlands',
                      'Norway', 'Poland', 'Portugal', 'Spain', 'Sweden', 'Switzerland', 'United Kingdom']

filtered_XmasAndNY = data[(data['Month'] == 12) & (data['Country'].isin(european_countries))]

# Create a binary DataFrame
basket = (filtered_XmasAndNY.groupby(['InvoiceNo', 'StockCode'])['Quantity']
          .sum().unstack().reset_index().fillna(0)
          .set_index('InvoiceNo'))

basket_sets = basket.applymap(lambda x: 1 if x > 0 else 0)

# Run Apriori algorithm to find frequent item sets
frequent_itemsets = apriori(basket_sets, min_support=0.03, use_colnames=True)

# Display frequent item sets with more than one item
for itemset in frequent_itemsets[frequent_itemsets['itemsets'].apply(lambda x: len(x) > 1)].itertuples(index=False):
    print('yes')
    print(itemset)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
/usr/local/lib/python3.10/dist-packages/mlxtend/frequent_patterns/fpcommon.py:110: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


yes
Pandas(support=0.04790221341261976, itemsets=frozenset({'22086', '22910'}))
yes
Pandas(support=0.03303600925008259, itemsets=frozenset({'22112', '22111'}))
yes
Pandas(support=0.036669970267591674, itemsets=frozenset({'22112', '22835'}))
yes
Pandas(support=0.036339610175090846, itemsets=frozenset({'22726', '22727'}))
yes
Pandas(support=0.030393128510075983, itemsets=frozenset({'22865', '22866'}))
yes
Pandas(support=0.030393128510075983, itemsets=frozenset({'22865', '22867'}))


In [16]:
# Coding find Product(Descripyion) with StockCode
def Product(x,y) :
  item =set(data[(data['StockCode']==x)|(data['StockCode']==y)]['Description'])
  return  item

# Create a summary table for comparing two products
product_pairs = [('22086', '22910'), ('22112', '22111'), ('22835', '22112'),
                 ('22726', '22727'), ('22865', '22866'), ('22865', '22867')]

# Create a DataFrame from the summary data
summary_df = pd.DataFrame(columns=['Product Pair', 'Support', 'Product 1 Description', 'Product 2 Description'])

# Populate the DataFrame
for product_pair in product_pairs:
    support = frequent_itemsets[
        frequent_itemsets['itemsets'] == frozenset(product_pair)]['support'].values[0]
    descriptions = Product(*product_pair)

    summary_df = summary_df.append({
        'Product Pair': product_pair,
        'Support': support,
        'Product 1 Description': descriptions.pop(),
        'Product 2 Description': descriptions.pop()
    }, ignore_index=True)

# Display the summary DataFrame
summary_df

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)
<ipython-input-16-32d461a366fc>:14: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  summary_df = summary_df.append({
<ipython-input-16-32d461a366fc>:14: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  summary_df = summary_df.append({
<ipython-input-16-32d461a366fc>:14: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  summary_df = summary_df.append({
<ipython-input-16-32d461a366fc>:1

Product Pair   Support            Product 1 Description  \
0  (22086, 22910)  0.047902  PAPER CHAIN KIT 50'S CHRISTMAS    
1  (22112, 22111)  0.033036       CHOCOLATE HOT WATER BOTTLE   
2  (22835, 22112)  0.036670       CHOCOLATE HOT WATER BOTTLE   
3  (22726, 22727)  0.036340        ALARM CLOCK BAKELIKE RED    
4  (22865, 22866)  0.030393                              NaN   
5  (22865, 22867)  0.030393          HAND WARMER BIRD DESIGN   

               Product 2 Description  
0  PAPER CHAIN KIT VINTAGE CHRISTMAS  
1       SCOTTIE DOG HOT WATER BOTTLE  
2    HOT WATER BOTTLE I AM SO POORLY  
3         ALARM CLOCK BAKELIKE GREEN  
4             HAND WARMER OWL DESIGN  
5                                NaN

##5.ของวันวาเลนไทน

##6.โดยจัดกลุ่มตามประเทศ+เดือน